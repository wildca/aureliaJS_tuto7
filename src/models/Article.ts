export interface Article {
    readonly author: string;
    readonly feedId: number;
    readonly feedTitle: string;
    readonly id: number;
    readonly isRead: boolean;
    readonly isFavorite: boolean;
    readonly link: string;
    readonly title: string;
    readonly updatedAt: Date;
}
