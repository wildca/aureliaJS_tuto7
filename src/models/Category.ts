export interface Category {
    id: string;
    title: string;
    unreadCount: number;
}
