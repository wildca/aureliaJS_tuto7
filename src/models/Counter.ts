export interface Counter {
    categoryId: string;
    unreadCount: number;
}
