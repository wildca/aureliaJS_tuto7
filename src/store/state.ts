import environment from '../environment';
import {Article} from '../models/Article';
import {Category} from '../models/Category';

/**
 * Data from this interface is restored from localStorage. Please dont introduce breaking changes.
 */
export interface Auth {
    host: string;
    isLoggedIn: boolean;
    password: string;
    token: string;
    username: string;
}

export enum SelectableBackend {
    ttrss = 'TTRSS',
    test = 'test',
}

export interface Route {
    name: string;
    params: object;
}

export interface Router {
    currentRoute: Route;
    newRoute: Route | null;
}

export interface RSS {
    isFetching: boolean;
    categories: Category[];
    currentCategoryId: string;
    // If an article is in this array, instead of marking it as read, we remove it from the list
    // of displayed articles.
    articlesToDiscards: Article[];
    displayedArticles: Article[];
}

export interface State {
    authentication: Auth;
    isOnline: boolean;
    router: Router;
    rss: RSS;
    selectableBackends: SelectableBackend[];
    selectedBackend: SelectableBackend;
}

const selectableBackends: SelectableBackend[] = [
    SelectableBackend.ttrss,
];

if (environment.debug) {
    selectableBackends.push(SelectableBackend.test);
}

export const initialState: State = {
    authentication: {
        host: '',
        isLoggedIn: false,
        password: '',
        username: '',
        token: '',
    },
    isOnline: true,
    router: {
        currentRoute: {
            name: '',
            params: {},
        },
        newRoute: null,
    },
    rss: {
        isFetching: false,
        categories: [],
        currentCategoryId: '',
        articlesToDiscards: [],
        displayedArticles: [],
    },
    selectableBackends,
    // Use the first backend of the list as the default backend.
    selectedBackend: selectableBackends[0],
};
