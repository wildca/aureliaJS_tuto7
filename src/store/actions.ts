import {State} from './state';

export function doRouterNavigation(state: State, newRouteName: string, newRouteParams: any = {}) {
    const newState = {...state};
    newState.router = {...state.router};
    newState.router.newRoute = {
        name: newRouteName,
        params: newRouteParams,
    };
    return newState;
}

export function isOffline(state: State) {
    const newState = {...state};
    newState.isOnline = false;
    return newState;
}

export function isOnline(state: State) {
    const newState = {...state};
    newState.isOnline = true;
    return newState;
}

export function logout(state: State) {
    const newState = {...state};
    newState.rss = {...state.rss};
    newState.authentication = {
        host: '',
        isLoggedIn: false,
        username: '',
        password: '',
        token: '',
    };
    newState.rss.articlesToDiscards = [];
    newState.rss.displayedArticles = [];
    newState.rss.categories = [];
    return doRouterNavigation(newState, 'login');
}

export function routerPerformedNavigation(
    state: State, currentRouteName: string, currentRouteParams,
) {
    const newState = {...state};
    newState.router = {...state.router};
    newState.rss = {...state.rss};
    newState.router.newRoute = null;
    newState.router.currentRoute = {
        name: currentRouteName,
        params: currentRouteParams,
    };

    if (currentRouteParams && currentRouteParams.categoryId) {
        newState.rss.currentCategoryId = currentRouteParams.categoryId.toString();
    } else {
        newState.rss.currentCategoryId = '';
    }

    return newState;
}
