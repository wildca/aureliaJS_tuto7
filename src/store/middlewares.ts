import {rehydrateFromLocalStorage} from 'aurelia-store';
import {State} from './state';

export function rehydratePartiallyFromLocalStorage(state: State, key?: string) {
    const stateFromLocalStorage: State = rehydrateFromLocalStorage(state, key);
    const newState = {...state};
    if (stateFromLocalStorage.authentication) {
        newState.authentication = stateFromLocalStorage.authentication;
    }

    return newState;
}
