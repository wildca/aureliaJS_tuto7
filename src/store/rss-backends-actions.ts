import {Container} from 'aurelia-framework';
import * as LogManager from 'aurelia-logging';
import {Store} from 'aurelia-store';
import * as constants from '../constants';
import {Article} from '../models/Article';
import {Category} from '../models/Category';
import {Counter} from '../models/Counter';
import {Backend} from '../services/backend';
import {TestBackend} from '../services/backends/test';
import {TTRss} from '../services/backends/ttrss';
import {doRouterNavigation} from './actions';
import {Auth, SelectableBackend, State} from './state';

const logger = LogManager.getLogger('aurss:store:actions:rss');

export function selectBackend(state: State, selectedBackend: SelectableBackend) {
    const newState = {...state};
    newState.selectedBackend = selectedBackend;
    return newState;
}

/**
 * This action is meant to be called when the authenticate method of the backend succeeds.
 * @param state
 * @param authInfos
 */

export function authenticateSucceeded(state: State, authInfos: Auth) {
    const newState = Object.assign({}, state);
    newState.authentication = authInfos;
    return doRouterNavigation(
        newState, 'display-articles-from-category', { categoryId: constants.DEFAULT_CATEGORY},
    );
}

/**
 * This action is meant to be called when the isLoggedIn method of the backend returns true.
 * @param state
 */
export function isLoggedIn(state: State) {
    const newState = {...state};
    newState.authentication = {...state.authentication};
    newState.authentication.isLoggedIn = true;
    return newState;
}

export function fetchArticles(state: State, categoryId: string) {
    const newState = {...state};
    newState.rss = {...state.rss};
    newState.rss.isFetching = true;

    getBackend(state.selectedBackend).getArticles(categoryId);

    return newState;
}

export function receivedArticles(state: State, articles: Article[]) {
    const newState = {...state};
    newState.rss = {...state.rss};
    newState.rss.isFetching = false;
    newState.rss.displayedArticles = articles;

    return newState;
}

export function fetchCategories(state: State) {
    getBackend(state.selectedBackend).getCategories();
    return state;
}

export function receivedCategories(state: State, categories: Category[]) {
    const newState = {...state};
    newState.rss = {...state.rss};
    newState.rss.categories = categories;

    return newState;
}

export function fetchCounters(state: State) {
    getBackend(state.selectedBackend).getCounters();
    return state;
}

export function receivedCounters(state: State, counters: Counter[]) {
    const newState = {...state};
    newState.rss = {...state.rss};

    const categoryIdToCounters = counters.reduce((acc, counter) => {
        acc[counter.categoryId] = counter.unreadCount;
        return acc;
    }, {});

    newState.rss.categories = state.rss.categories.map((category) => ({
        ...category,
        unreadCount: categoryIdToCounters[category.id],
    }));

    return newState;
}

/**
 * Do a request to the backend to mark an article as read.
 * @param state
 * @param article
 * @param discard: If this is true, the article will be added to articlesToDiscards.
 */
export function markAsRead(state: State, article: Article, { discard = false } = {}) {
    getBackend(state.selectedBackend).markAsRead(article);
    if (!discard) {
        return state;
    }
    const newState = {...state};
    newState.rss = {...newState.rss};
    newState.rss.articlesToDiscards = [
        ...newState.rss.articlesToDiscards,
        article,
    ];

    return newState;
}

export function markAsUnread(state: State, article: Article) {
    getBackend(state.selectedBackend).markAsUnread(article);
    return state;
}

export function markAsFavorite(state: State, article: Article) {
    getBackend(state.selectedBackend).markAsFavorite(article);
    return state;
}

export function unmarkAsFavorite(state: State, article: Article) {
    getBackend(state.selectedBackend).unmarkAsFavorite(article);
    return state;
}

export function markedAsRead(state: State, article: Article) {
    fetchCounters(state);
    if (state.rss.articlesToDiscards.includes(article)) {
        return removeArticle(state, article);
    }

    return updateArticle(state, article, { isRead: true });
}

function removeArticle(state: State, article: Article) {
    if (!state.rss.displayedArticles.includes(article)) {
        return state;
    }

    const newState = {...state};
    newState.rss = {...state.rss};
    newState.rss.displayedArticles = [...newState.rss.displayedArticles];
    const indexToRemove = newState.rss.displayedArticles.indexOf(article);
    newState.rss.displayedArticles.splice(indexToRemove, 1);

    return newState;
}

function updateArticle(state: State, article: Article, patch) {
    const newState = {...state};
    newState.rss = {...newState.rss};

    newState.rss.displayedArticles = updateArticleInArray(
        article, newState.rss.displayedArticles, patch,
    );

    return newState;
}

function updateArticleInArray(article: Article, array: Article[], patch): Article[] {
    if (!array.includes(article)) {
        return array;
    }

    const newArray = [...array];
    const indexUpdatedArticle = newArray.indexOf(article);
    newArray[indexUpdatedArticle] = {
        ...newArray[indexUpdatedArticle],
        ...patch,
    };
    return newArray;
}

export function markedAsUnread(state: State, article: Article) {
    fetchCounters(state);
    return updateArticle(state, article, { isRead: false });
}

export function markedAsFavorite(state: State, article: Article) {
    return updateArticle(state, article, { isFavorite: true });
}

export function unmarkedAsFavorite(state: State, article: Article) {
    return updateArticle(state, article, { isFavorite: false });
}

export function openArticle(state: State, article: Article) {
    const tab = window.open(article.link, '_blank');
    if (tab) {
        tab.focus();
    }

    return state;
}

export function getBackend(selectedBackend: SelectableBackend): Backend {
    switch (selectedBackend) {
        case SelectableBackend.ttrss:
            return Container.instance.get(TTRss) as Backend;
        case SelectableBackend.test:
            return Container.instance.get(TestBackend) as Backend;
        default:
            logger.error('An unsupported backend was requested, return the default test backend');
            return Container.instance.get(TestBackend) as Backend;
    }
}

// This must be exported because to tests some component that uses dispatchify,
// these actions must be registered.
export function registerRssBackendsActions(store: Store<State>) {
    store.registerAction(
        selectBackend.name,
        selectBackend,
    );
    store.registerAction(
        authenticateSucceeded.name,
        authenticateSucceeded,
    );
    store.registerAction(
        isLoggedIn.name,
        isLoggedIn,
    );
    store.registerAction(
        markAsRead.name,
        markAsRead,
    );
    store.registerAction(
        markAsUnread.name,
        markAsUnread,
    );
    store.registerAction(
        markedAsRead.name,
        markedAsRead,
    );
    store.registerAction(
        markAsFavorite.name,
        markAsFavorite,
    );
    store.registerAction(
        unmarkAsFavorite.name,
        unmarkAsFavorite,
    );
    store.registerAction(
        markedAsUnread.name,
        markedAsUnread,
    );
    store.registerAction(
        markedAsFavorite.name,
        markedAsFavorite,
    );
    store.registerAction(
        unmarkedAsFavorite.name,
        unmarkedAsFavorite,
    );
    store.registerAction(
        openArticle.name,
        openArticle,
    );
    store.registerAction(
        fetchArticles.name,
        fetchArticles,
    );
    store.registerAction(
        receivedArticles.name,
        receivedArticles,
    );
    store.registerAction(
        fetchCategories.name,
        fetchCategories,
    );
    store.registerAction(
        receivedCategories.name,
        receivedCategories,
    );
    store.registerAction(
        fetchCounters.name,
        fetchCounters,
    );
    store.registerAction(
        receivedCounters.name,
        receivedCounters,
    );
}
