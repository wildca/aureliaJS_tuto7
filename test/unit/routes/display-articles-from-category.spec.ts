import {bootstrap} from 'aurelia-bootstrapper';
import {Container} from 'aurelia-framework';
import {PLATFORM} from 'aurelia-pal';
import {Store} from 'aurelia-store';
import {StageComponent} from 'aurelia-testing';
import {Article} from '../../../src/models/Article';
import {receivedArticles, registerRssBackendsActions} from '../../../src/store/rss-backends-actions';
import {createInitialState} from '../../__fixtures__/state-fixtures';

describe('Route: DisplayArticlesFromCategory', () => {
    describe('Render tests', () => {
        let component;
        let store;
        let state;

        beforeEach(async() => {
            state = createInitialState();
            state.rss.isFetching = true;
            store = new Store(state);
            registerRssBackendsActions(store);
            Container.instance.registerInstance(Store, store);

            component = StageComponent.withResources(PLATFORM.moduleName('../../src/routes/display-articles-from-category'))
                .inView('<display-articles-from-category></display-articles-from-category>');

            await component.create(bootstrap);
            component.viewModel.store = store;
            component.viewModel.activate();

            return component;
        });

        afterEach(() => {
            component.dispose();
        });

        it('should render and match snapshot while fetching', async() => {
            await component.waitForElement('.fa-spinner');

            expect(document.body.outerHTML).toMatchSnapshot();
        });

        it('should render and match snapshot after fetching and no result', async() => {
            await store.dispatch(receivedArticles, []);
            await component.waitForElement('.articles-container');

            expect(document.body.outerHTML).toMatchSnapshot();
        });

        it('should render and match snapshot after fetching and results', async() => {
            const articles: Article[] = [{
                author: 'Tester',
                title: 'Test article',
                feedId: 0,
                feedTitle: 'Test feed',
                id: 1,
                isRead: false,
                isFavorite: false,
                updatedAt: new Date(),
                link: 'https://example.com',
            }];
            await store.dispatch(receivedArticles, articles);
            await component.waitForElements('aurss-article');

            expect(document.body.outerHTML).toMatchSnapshot();
        });
    });
});
