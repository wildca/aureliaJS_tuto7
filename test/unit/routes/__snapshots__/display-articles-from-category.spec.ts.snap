// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Route: DisplayArticlesFromCategory Render tests should render and match snapshot after fetching and no result 1`] = `
"<body><div><display-articles-from-category class=\\"au-target\\" au-target-id=\\"8\\">
    <h2 class=\\"subtitle is-2\\"></h2>

    <!--anchor-->
    <div class=\\"articles-container\\">
        <p t=\\"display-articles-from-category-no-article\\">No article.</p><!--anchor-->

        <div class=\\"tile is-ancestor is-vertical\\">
            <!--anchor-->
        </div>
    </div><!--anchor-->
</display-articles-from-category></div></body>"
`;

exports[`Route: DisplayArticlesFromCategory Render tests should render and match snapshot after fetching and results 1`] = `
"<body><div><display-articles-from-category class=\\"au-target\\" au-target-id=\\"9\\">
    <h2 class=\\"subtitle is-2\\"></h2>

    <!--anchor-->
    <div class=\\"articles-container\\">
        <!--anchor-->

        <div class=\\"tile is-ancestor is-vertical\\">
            <aurss-article aurss-swipable=\\"on-swipe-success.bind: markArticleAsReadAndDiscard; on-swipe-success-arg.bind: article\\" aurss-mark-as-read-on-scroll=\\"mark-as-read.bind: markArticleAsRead; article.bind: article\\" value.one-time=\\"article\\" mark-article-as-read.one-time=\\"markArticleAsRead\\" mark-article-as-unread.one-time=\\"markArticleAsUnread\\" mark-article-as-favorite.one-time=\\"markArticleAsFavorite\\" unmark-article-as-favorite.one-time=\\"unmarkArticleAsFavorite\\" open-article.one-time=\\"openArticle\\" class=\\"au-target\\" au-target-id=\\"4\\"></aurss-article><!--anchor-->
        </div>
    </div><!--anchor-->
</display-articles-from-category></div></body>"
`;

exports[`Route: DisplayArticlesFromCategory Render tests should render and match snapshot while fetching 1`] = `
"<body><div><display-articles-from-category class=\\"au-target\\" au-target-id=\\"7\\">
    <h2 class=\\"subtitle is-2\\"></h2>

    <div class=\\"has-text-centered\\">
        <i class=\\"fas fa-spinner fa-spin is-size-1\\"></i>
    </div><!--anchor-->
    <!--anchor-->
</display-articles-from-category></div></body>"
`;
